File: observability/src/commonMain/kotlin/dev/readthat/observability/ProductAnalytics.kt
package dev.readthat.observability

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.encodeToJsonElement

/** Bounded product-event names. Content identifiers belong in [ProductEvent.contentId]. */
object ProductEventName {
    const val SESSION_START = "session_start"
    const val SESSION_FOREGROUND = "session_foreground"
    const val SESSION_CHECKPOINT = "session_checkpoint"
    const val SESSION_SUMMARY = "session_summary"
    const val POST_IMPRESSION = "post_impression"
    const val POST_DETAIL_VIEW = "post_detail_view"
    const val COMMENTS_VIEW = "comments_view"
    const val COMMENT_CREATE = "comment_create"
    const val MEDIA_PLAYBACK = "media_playback"
    const val MEDIA_FEED_TIME_SPENT = "media_feed_time_spent"
    const val COMMUNITY_VIEW = "community_view"
    const val COMMUNITY_TIME_SPENT = "community_time_spent"
    const val COMMUNITY_JOIN = "community_join"
    const val COMMUNITY_LEAVE = "community_leave"
    const val COMMUNITY_POST_VIEW = "community_post_view"
    const val AD_IMPRESSION = "ad_impression"
    const val AD_VIEW_TIME = "ad_view_time"
    const val AD_CLICK = "ad_click"
    const val AD_CTA_CLICK = "ad_cta_click"
    const val AD_CAROUSEL_SWIPE = "ad_carousel_swipe"
    const val AD_RELATED_CLICK = "ad_related_click"
    const val AD_VIDEO_PLAY = "ad_video_play"
    const val AD_VIDEO_WATCH = "ad_video_watch"
    const val AD_VIDEO_COMPLETE = "ad_video_complete"
    const val AD_DETAIL_VIEW = "ad_detail_view"
    const val AD_LANDING_LOAD = "ad_landing_load"
}

@Serializable
enum class ProductSurface { APP, FEED, DETAIL, COMMENTS, COMMUNITY, MEDIA, AD_DETAIL, BACKGROUND }

@Serializable
enum class ProductContentType { POST, COMMENT, COMMUNITY, VIDEO, AD }

@Serializable
enum class ProductEventReason {
    COLD_START,
    FOREGROUND,
    BACKGROUND,
    TIMEOUT,
    IDENTITY_CHANGE,
    PAUSE,
    ENDED,
    MEDIA_CHANGE,
    SURFACE_CHANGE,
    ERROR,
}

/**
 * Privacy-bounded behavior event. The server irreversibly pseudonymizes
 * [contentId] before storage; no titles, bodies, URLs, usernames, or tokens are
 * accepted by the wire schema.
 */
@Serializable
data class ProductEvent(
    val name: String,
    val surface: ProductSurface,
    val recordedAtEpochMs: Long = epochMilliseconds(),
    val contentId: String? = null,
    val contentType: ProductContentType? = null,
    val reason: ProductEventReason? = null,
    val durationMs: Long? = null,
    val position: Int? = null,
    val completionPercent: Double? = null,
)

@Serializable
data class ProductAnalyticsBatch(
    val schemaVersion: Int = 1,
    val platform: String,
    val appVersion: String,
    val buildType: String,
    /** Random UUID persisted only in app-private storage; never stored raw by the server. */
    val installationId: String,
    /** Foreground engagement session, distinct from both process and authentication sessions. */
    val sessionId: String,
    val events: List<ProductEvent>,
)

object ProductAnalyticsWireFormat {
    private val json = Json { explicitNulls = false; encodeDefaults = true }

    fun encode(batch: ProductAnalyticsBatch): JsonElement = json.encodeToJsonElement(batch)
}

fun interface ProductAnalyticsRecorder {
    /** Must return immediately; the platform exporter owns persistence and network work. */
    fun record(event: ProductEvent)
}

/** Vendor-neutral feature seam. The Android app installs its lifecycle-aware exporter. */
object ProductAnalytics : ProductAnalyticsRecorder {
    private var delegate: ProductAnalyticsRecorder = ProductAnalyticsRecorder { }

    fun install(recorder: ProductAnalyticsRecorder) {
        delegate = recorder
    }

    override fun record(event: ProductEvent) = delegate.record(event)
}

File: backend/src/product-analytics.ts
import { z } from "zod";
import { keyedHash } from "./crypto";
import { jsonResponse, readJson } from "./http";
import type { RequestContext } from "./types";

const eventName = z.enum([
  "session_start",
  "session_foreground",
  "session_checkpoint",
  "session_summary",
  "post_impression",
  "post_detail_view",
  "comments_view",
  "comment_create",
  "media_playback",
  "media_feed_time_spent",
  "community_view",
  "community_time_spent",
  "community_join",
  "community_leave",
  "community_post_view",
  "ad_impression",
  "ad_view_time",
  "ad_click",
  "ad_cta_click",
  "ad_carousel_swipe",
  "ad_related_click",
  "ad_video_play",
  "ad_video_watch",
  "ad_video_complete",
  "ad_detail_view",
  "ad_landing_load",
]);

const eventSchema = z.object({
  name: eventName,
  surface: z.enum(["APP", "FEED", "DETAIL", "COMMENTS", "COMMUNITY", "MEDIA", "AD_DETAIL", "BACKGROUND"]),
  recordedAtEpochMs: z.number().int().positive(),
  // Deliberately excludes slashes, spaces, and URL punctuation. The value is
  // HMACed below and is never written or logged in raw form.
  contentId: z.string().regex(/^[A-Za-z0-9._:-]{1,160}$/u).optional(),
  contentType: z.enum(["POST", "COMMENT", "COMMUNITY", "VIDEO", "AD"]).optional(),
  reason: z.enum([
    "COLD_START",
    "FOREGROUND",
    "BACKGROUND",
    "TIMEOUT",
    "IDENTITY_CHANGE",
    "PAUSE",
    "ENDED",
    "MEDIA_CHANGE",
    "SURFACE_CHANGE",
    "ERROR",
  ]).optional(),
  durationMs: z.number().int().min(0).max(7 * 24 * 60 * 60 * 1_000).optional(),
  position: z.number().int().min(0).max(24 * 60 * 60 * 1_000).optional(),
  completionPercent: z.number().finite().min(0).max(100).optional(),
}).strict();

const batchSchema = z.object({
  schemaVersion: z.literal(1),
  platform: z.enum(["android", "ios", "web"]),
  appVersion: z.string().regex(/^[A-Za-z0-9._+-]{1,32}$/u),
  buildType: z.enum(["debug", "release", "staging"]),
  installationId: z.string().uuid(),
  sessionId: z.string().uuid(),
  events: z.array(eventSchema).min(1).max(50),
}).strict();

async function pseudonym(secret: string, namespace: string, value: string): Promise<string> {
  return (await keyedHash(secret, `product-analytics:v1:${namespace}:${value}`)).slice(0, 32);
}

/**
 * Ingests bounded behavior events into a dataset separate from performance
 * distributions. index1 is a stable, server-only pseudonym for either the
 * authenticated user or installation; raw account/install/content ids are not stored.
 */
export async function ingestProductAnalytics(context: RequestContext): Promise<Response> {
  const batch = await readJson(context.request, batchSchema, 64 * 1024);
  const principalType = context.viewer ? "user" : "installation";
  const principalSource = context.viewer?.id ?? batch.installationId;
  const principalId = await pseudonym(
    context.env.ANALYTICS_ID_PEPPER,
    principalType,
    principalSource,
  );
  const colo = String(context.request.cf?.colo ?? "unknown");
  const country = String(context.request.cf?.country ?? "unknown");
  const contentKeys = new Map<string, string>();
  await Promise.all([...new Set(batch.events.map((event) => event.contentId).filter(
    (value): value is string => value !== undefined,
  ))].map(async (contentId) => {
    contentKeys.set(
      contentId,
      await pseudonym(context.env.ANALYTICS_ID_PEPPER, "content", contentId),
    );
  }));

  for (const event of batch.events) {
    const contentKey = event.contentId ? contentKeys.get(event.contentId) ?? "" : "";
    context.env.PRODUCT_ANALYTICS.writeDataPoint({
      indexes: [principalId],
      blobs: [
        event.name,
        batch.platform,
        batch.appVersion,
        batch.buildType,
        batch.sessionId,
        event.surface,
        contentKey,
        event.contentType ?? "",
        event.reason ?? "",
        principalType,
        colo,
        country,
      ],
      doubles: [
        event.recordedAtEpochMs,
        event.durationMs ?? 0,
        event.position ?? 0,
        event.completionPercent ?? 0,
      ],
    });
  }

  console.log(JSON.stringify({
    level: "info",
    message: "client product analytics batch accepted",
    requestId: context.requestId,
    platform: batch.platform,
    principalType,
    eventCount: batch.events.length,
  }));
  return jsonResponse(
    { accepted: batch.events.length, requestId: context.requestId },
    { status: 202 },
  );
}
